Unable to package windows with Docker

If I run the windows packaging script in docker, it fails with the below error.  Running the exact same logic locally works.  I am specifying SECRETS_DIR and BNS_CERT_PASS environment variables, and the SECRETS_DIR contains bns.pfx (not a symlink, but the actual file).  Note - I previously ran without that file, and got a different error (unable to open file or something like that).

```
Building windows/386...
Using /usr/local/bin/docker-machine
Build tags:
Extra ldflags: -X=github.com/getlantern/flashlight.compileTimePackageVersion=2.0.17
Generating distribution package for windows/386...
Using /usr/local/bin/docker-machine
Build tags:
Extra ldflags: -X=github.com/getlantern/flashlight.compileTimePackageVersion=2.0.17
/bin/bash: line 15:   898 Killed                  osslsigncode sign -pkcs12 "$BNS_CERT" -pass "$BNS_CERT_PASS" -n "Lantern" -t http://timestamp.verisign.com/scripts/timstamp.dll -in "lantern_windows_386.exe" -out "$INSTALLER_RESOURCES/lantern.exe"
Makefile:238: recipe for target 'docker-package-windows' failed
make: *** [docker-package-windows] Error 137
make: *** [package-windows] Error 2
```

--- a/.gitignore
+++ b/.gitignore
@@ -57,3 +57,4 @@ local.properties
 build
 keystore.release.jks
 src/github.com/getlantern/lantern-ui/app/js/revision.js
+bns.pfx

--- a/Makefile
+++ b/Makefile
@@ -563,6 +563,9 @@ android-install: $(LANTERN_MOBILE_ANDROID_DEBUG)
 clean-assets:
 	rm -f $(RESOURCES_DOT_GO)
 
+# Provided for backward compatibility with how people used to use the makefile
+update-dist: clean-assets assets
+
 clean-bns-cert:
 	rm -f $(BNS_CERT)
 

--- a/Makefile
+++ b/Makefile
@@ -35,6 +35,9 @@ PACKAGE_URL := https://www.getlantern.org
 PACKAGED_YAML := .packaged-lantern.yaml
 MANOTO_YAML := .packaged-lantern-manoto.yaml
 
+RESOURCES_DOT_GO := ./src/github.com/getlantern/flashlight/ui/resources.go
+BNS_CERT := bns.pfx
+
 LANTERN_BINARIES_PATH ?= ../lantern-binaries
 
 GH_USER ?= getlantern
@@ -94,29 +97,6 @@ define build-tags
 	EXTRA_LDFLAGS=$$(echo $$EXTRA_LDFLAGS | xargs) && echo "Extra ldflags: $$EXTRA_LDFLAGS"
 endef
 
-define docker-up
-	if [[ "$$(uname -s)" == "Darwin" ]]; then \
-		if [[ -z "$(DOCKERMACHINE)" ]]; then \
-		  if [[ -z "$(BOOT2DOCKER)" ]]; then \
-  			echo 'Missing "docker-machine" command' && exit 1; \
-			fi && \
-			echo "Falling back to using $(BOOT2DOCKER), recommend upgrading to latest docker toolbox from https://www.docker.com/docker-toolbox" && \
-			if [[ "$$($(BOOT2DOCKER) status)" != "running" ]]; then \
-				$(BOOT2DOCKER) up; \
-			fi && \
-			if [[ -z "$$DOCKER_HOST" ]]; then \
-				$$($(BOOT2DOCKER) shellinit 2>/dev/null); \
-			fi \
-		else \
-		  echo "Using $(DOCKERMACHINE)" && \
-			if [[ "$$($(DOCKERMACHINE) status default)" != "Running" ]]; then \
-				$(DOCKERMACHINE) start default; \
-			fi && \
-			$$($(DOCKERMACHINE) env default 2>/dev/null | head -4 | tr -d '"'); \
-		fi \
-	fi
-endef
-
 define fpm-debian-build =
 	echo "Running fpm-debian-build" && \
  	PKG_ARCH=$1 && \
@@ -153,59 +133,84 @@ define fpm-debian-build =
 	fpm -a $$PKG_ARCH -s dir -t deb -n lantern-manoto -v $$VERSION -m "$(PACKAGE_MAINTAINER)" --description "$(LANTERN_DESCRIPTION)\n$(LANTERN_EXTENDED_DESCRIPTION)" --category net --license "Apache-2.0" --vendor "$(PACKAGE_VENDOR)" --url $(PACKAGE_URL) --deb-compression xz -f -C $$WORKDIR usr;
 endef
 
+define docker-up
+	if [[ "$$(uname -s)" == "Darwin" ]]; then \
+		if [[ -z "$(DOCKERMACHINE)" ]]; then \
+		  if [[ -z "$(BOOT2DOCKER)" ]]; then \
+  			echo 'Missing "docker-machine" command' && exit 1; \
+			fi && \
+			echo "Falling back to using $(BOOT2DOCKER), recommend upgrading to latest docker toolbox from https://www.docker.com/docker-toolbox" && \
+			if [[ "$$($(BOOT2DOCKER) status)" != "running" ]]; then \
+				$(BOOT2DOCKER) up; \
+			fi && \
+			if [[ -z "$$DOCKER_HOST" ]]; then \
+				$$($(BOOT2DOCKER) shellinit 2>/dev/null); \
+			fi \
+		else \
+		  echo "Using $(DOCKERMACHINE)" && \
+			if [[ "$$($(DOCKERMACHINE) status default)" != "Running" ]]; then \
+				$(DOCKERMACHINE) start default; \
+			fi && \
+			$$($(DOCKERMACHINE) env default 2>/dev/null | head -4 | tr -d '"'); \
+		fi \
+	fi
+endef
+
+# This implicit rule allows prefix any existing target with "docker-" to make it
+# run in docker.
+docker-%: system-checks
+	@$(call docker-up) && \
+	DOCKER_CONTEXT=.$(DOCKER_IMAGE_TAG)-context && \
+	mkdir -p $$DOCKER_CONTEXT && \
+	cp Dockerfile $$DOCKER_CONTEXT && \
+	docker build -t $(DOCKER_IMAGE_TAG) $$DOCKER_CONTEXT && \
+	docker run -v $$PWD:/lantern -t $(DOCKER_IMAGE_TAG) /bin/bash -c 'cd /lantern && VERSION="'$$VERSION'" HEADLESS="'$$HEADLESS'" SECRETS_DIR="'$$SECRETS_DIR'" BNS_CERT_PASS="'$$BNS_CERT_PASS'" make $*';
+
 all: binaries
 android-dist: genconfig android
 
-# This is to be called within the docker image.
-docker-genassets: require-npm
+$(RESOURCES_DOT_GO): $(NPM)
 	@source setenv.bash && \
 	LANTERN_UI="src/github.com/getlantern/lantern-ui" && \
 	APP="$$LANTERN_UI/app" && \
 	DIST="$$LANTERN_UI/dist" && \
 	echo 'var LANTERN_BUILD_REVISION = "$(GIT_REVISION_SHORTCODE)";' > $$APP/js/revision.js && \
 	git update-index --assume-unchanged $$APP/js/revision.js && \
-	if [[ ! -d $$DIST ]]; then \
-		UPDATE_DIST=true; \
-	fi && \
-	DEST="src/github.com/getlantern/flashlight/ui/resources.go" && \
-	\
-	if [ "$$UPDATE_DIST" ]; then \
-			cd $$LANTERN_UI && \
-			npm install && \
-			rm -Rf dist && \
-			gulp build && \
-			cd -; \
-	fi && \
+	DEST="$@" && \
+	cd $$LANTERN_UI && \
+	npm install && \
+	rm -Rf dist && \
+	gulp build && \
+	cd - && \
 	rm -f bin/tarfs bin/rsrc && \
 	go install github.com/getlantern/tarfs/tarfs && \
 	echo "// +build !stub" > $$DEST && \
 	echo " " >> $$DEST && \
 	tarfs -pkg ui $$DIST >> $$DEST && \
 	go install github.com/akavel/rsrc && \
-	rsrc -ico installer-resources/windows/lantern.ico -o src/github.com/getlantern/flashlight/lantern_windows_386.syso;
+	rsrc -ico installer-resources/windows/lantern.ico -o src/github.com/getlantern/flashlight/lantern_windows_386.syso
 
-docker-linux-386:
+assets: $(RESOURCES_DOT_GO)
+
+linux-386: $(RESOURCES_DOT_GO)
 	@source setenv.bash && \
 	$(call build-tags) && \
 	CGO_ENABLED=1 GOOS=linux GOARCH=386 go build -a -o lantern_linux_386 -tags="$$BUILD_TAGS" -ldflags="$(LDFLAGS) $$EXTRA_LDFLAGS -linkmode internal -extldflags \"-static\"" github.com/getlantern/flashlight/main
 
-docker-linux-amd64:
+linux-amd64: $(RESOURCES_DOT_GO)
 	@source setenv.bash && \
 	$(call build-tags) && \
 	CGO_ENABLED=1 GOOS=linux GOARCH=amd64 go build -a -o lantern_linux_amd64 -tags="$$BUILD_TAGS" -ldflags="$(LDFLAGS) $$EXTRA_LDFLAGS -linkmode internal -extldflags \"-static\"" github.com/getlantern/flashlight/main
 
-docker-linux-arm:
+linux-arm: $(RESOURCES_DOT_GO)
 	@source setenv.bash && \
 	$(call build-tags) && \
 	CGO_ENABLED=1 CC=arm-linux-gnueabi-gcc CXX=arm-linux-gnueabi-g++ CGO_ENABLED=1 GOOS=linux GOARCH=arm GOARM=7 go build -a -o lantern_linux_arm -tags="$$BUILD_TAGS" -ldflags="$(LDFLAGS) $$EXTRA_LDFLAGS -linkmode internal -extldflags \"-static\"" github.com/getlantern/flashlight/main
 
-docker-windows-386:
+windows: $(RESOURCES_DOT_GO)
 	@source setenv.bash && \
 	$(call build-tags) && \
 	CGO_ENABLED=1 GOOS=windows GOARCH=386 go build -a -o lantern_windows_386.exe -tags="$$BUILD_TAGS" -ldflags="$(LDFLAGS) $$EXTRA_LDFLAGS -H=windowsgui" github.com/getlantern/flashlight/main;
-
-require-assets:
-	@if [ ! -f ./src/github.com/getlantern/flashlight/ui/resources.go ]; then make genassets; fi
 
 require-version:
 	@if [[ -z "$$VERSION" ]]; then echo "VERSION environment value is required."; exit 1; fi
@@ -214,10 +219,13 @@ require-gh-token:
 	@if [[ -z "$$GH_TOKEN" ]]; then echo "GH_TOKEN environment value is required."; exit 1; fi
 
 require-secrets-dir:
-	if [[ -z "$$SECRETS_DIR" ]]; then echo "SECRETS_DIR environment value is required."; exit 1; fi
+	@if [[ -z "$$SECRETS_DIR" ]]; then echo "SECRETS_DIR environment value is required."; exit 1; fi
 
 require-secrets: require-secrets-dir
 	@if [[ -z "$$BNS_CERT_PASS" ]]; then echo "BNS_CERT_PASS environment value is required."; exit 1; fi
+
+$(BNS_CERT): require-secrets-dir
+	@cp -f $$SECRETS_DIR/$(BNS_CERT) $(BNS_CERT)
 
 require-lantern-binaries:
 	@if [[ ! -d "$(LANTERN_BINARIES_PATH)" ]]; then \
@@ -225,54 +233,47 @@ require-lantern-binaries:
 		exit 1; \
 	fi
 
-docker-package-debian-386: require-version docker-linux-386
+package-linux-386: require-version linux-386
 	@cp lantern_linux_386 lantern_linux_i386;
 	@$(call fpm-debian-build,"i386")
 	@rm lantern_linux_i386;
 	@mv update_linux_i386.bz2 update_linux_386.bz2 && \
 	echo "-> lantern_$(VERSION)_i386.deb"
 
-docker-package-debian-amd64: require-version docker-linux-amd64
+package-linux-amd64: require-version linux-amd64
 	@$(call fpm-debian-build,"amd64")
 	@echo "-> lantern_$(VERSION)_amd64.deb"
 
-docker-package-debian-arm: require-version docker-linux-arm
+package-linux-arm: require-version linux-arm
 	@$(call fpm-debian-build,"arm")
 	@echo "-> lantern_$(VERSION)_arm.deb"
 
-docker-package-windows: require-version docker-windows-386
-	@if [[ -z "$$BNS_CERT" ]]; then echo "BNS_CERT environment value is required."; exit 1; fi && \
-	if [[ -z "$$BNS_CERT_PASS" ]]; then echo "BNS_CERT_PASS environment value is required."; exit 1; fi && \
+package-windows: $(BNS_CERT) do-package-windows
+
+do-package-windows: require-version windows
+	@if [[ -z "$$BNS_CERT_PASS" ]]; then echo "BNS_CERT_PASS environment value is required."; exit 1; fi && \
 	INSTALLER_RESOURCES="installer-resources/windows" && \
 	rm -f $$INSTALLER_RESOURCES/$(PACKAGED_YAML) && \
 	rm -f $$INSTALLER_RESOURCES/$(LANTERN_YAML) && \
 	cp installer-resources/$(PACKAGED_YAML) $$INSTALLER_RESOURCES/$(PACKAGED_YAML) && \
 	cp $(LANTERN_YAML_PATH) $$INSTALLER_RESOURCES/$(LANTERN_YAML) && \
-	osslsigncode sign -pkcs12 "$$BNS_CERT" -pass "$$BNS_CERT_PASS" -n "Lantern" -t http://timestamp.verisign.com/scripts/timstamp.dll -in "lantern_windows_386.exe" -out "$$INSTALLER_RESOURCES/lantern.exe" && \
+	osslsigncode sign -pkcs12 "$(BNS_CERT)" -pass "$$BNS_CERT_PASS" -n "Lantern" -t http://timestamp.verisign.com/scripts/timstamp.dll -in "lantern_windows_386.exe" -out "$$INSTALLER_RESOURCES/lantern.exe" && \
 	cat $$INSTALLER_RESOURCES/lantern.exe | bzip2 > update_windows_386.bz2 && \
 	ls -l lantern_windows_386.exe update_windows_386.bz2 && \
 	makensis -V1 -DVERSION=$$VERSION installer-resources/windows/lantern.nsi && \
-	osslsigncode sign -pkcs12 "$$BNS_CERT" -pass "$$BNS_CERT_PASS" -n "Lantern" -t http://timestamp.verisign.com/scripts/timstamp.dll -in "$$INSTALLER_RESOURCES/lantern-installer-unsigned.exe" -out "lantern-installer.exe" && \
+	osslsigncode sign -pkcs12 "$(BNS_CERT)" -pass "$$BNS_CERT_PASS" -n "Lantern" -t http://timestamp.verisign.com/scripts/timstamp.dll -in "$$INSTALLER_RESOURCES/lantern-installer-unsigned.exe" -out "lantern-installer.exe" && \
 	cp installer-resources/$(MANOTO_YAML) $$INSTALLER_RESOURCES/$(PACKAGED_YAML) && \
 	cp $(LANTERN_YAML_PATH) $$INSTALLER_RESOURCES/$(LANTERN_YAML) && \
 	makensis -V1 -DVERSION=$$VERSION installer-resources/windows/lantern.nsi && \
-	osslsigncode sign -pkcs12 "$$BNS_CERT" -pass "$$BNS_CERT_PASS" -n "Lantern" -t http://timestamp.verisign.com/scripts/timstamp.dll -in "$$INSTALLER_RESOURCES/lantern-installer-unsigned.exe" -out "lantern-installer-manoto.exe";
+	osslsigncode sign -pkcs12 "$(BNS_CERT)" -pass "$$BNS_CERT_PASS" -n "Lantern" -t http://timestamp.verisign.com/scripts/timstamp.dll -in "$$INSTALLER_RESOURCES/lantern-installer-unsigned.exe" -out "lantern-installer-manoto.exe" && \
+	echo "-> lantern-installer.exe and lantern-installer-manoto.exe"
 
-docker: system-checks
-	@$(call docker-up) && \
-	DOCKER_CONTEXT=.$(DOCKER_IMAGE_TAG)-context && \
-	mkdir -p $$DOCKER_CONTEXT && \
-	cp Dockerfile $$DOCKER_CONTEXT && \
-	docker build -t $(DOCKER_IMAGE_TAG) $$DOCKER_CONTEXT;
+# We overwrite the implicit docker-package-windows rule to include a dependency
+# on $(BNS_CERT)
+docker-package-windows: $(BNS_CERT)
+	make docker-do-package-windows
 
-update-dist:
-	UPDATE_DIST=true make genassets
-
-linux: genassets linux-386 linux-amd64
-
-windows: genassets windows-386
-
-darwin: genassets darwin-amd64
+linux: linux-386 linux-amd64
 
 system-checks:
 	@if [[ -z "$(DOCKER)" ]]; then echo 'Missing "docker" command.'; exit 1; fi && \
@@ -304,34 +305,7 @@ require-ruby:
 	(gem which octokit >/dev/null) || (echo 'Missing gem "octokit". Try sudo gem install octokit.' && exit 1) && \
 	(gem which mime-types >/dev/null) || (echo 'Missing gem "mime-types". Try sudo gem install mime-types.' && exit 1)
 
-genassets: docker
-	@echo "Generating assets..." && \
-	$(call docker-up) && \
-	docker run -v $$PWD:/lantern -t $(DOCKER_IMAGE_TAG) /bin/bash -c "cd /lantern && UPDATE_DIST=$$UPDATE_DIST make docker-genassets" && \
-	git update-index --assume-unchanged src/github.com/getlantern/flashlight/ui/resources.go && \
-	echo "OK"
-
-linux-386: require-assets docker
-	@echo "Building linux/386..." && \
-	$(call docker-up) && \
-	docker run -v $$PWD:/lantern -t $(DOCKER_IMAGE_TAG) /bin/bash -c 'cd /lantern && VERSION="'$$VERSION'" HEADLESS="'$$HEADLESS'" make docker-linux-386'
-
-linux-amd64: require-assets docker
-	@echo "Building linux/amd64..." && \
-	$(call docker-up) && \
-	docker run -v $$PWD:/lantern -t $(DOCKER_IMAGE_TAG) /bin/bash -c 'cd /lantern && VERSION="'$$VERSION'" HEADLESS="'$$HEADLESS'" make docker-linux-amd64'
-
-linux-arm: require-assets docker
-	@echo "Building linux/arm..." && \
-	$(call docker-up) && \
-	docker run -v $$PWD:/lantern -t $(DOCKER_IMAGE_TAG) /bin/bash -c 'cd /lantern && VERSION="'$$VERSION'" HEADLESS="1" make docker-linux-arm'
-
-windows-386: require-assets docker
-	@echo "Building windows/386..." && \
-	$(call docker-up) && \
-	docker run -v $$PWD:/lantern -t $(DOCKER_IMAGE_TAG) /bin/bash -c 'cd /lantern && VERSION="'$$VERSION'" HEADLESS="'$$HEADLESS'" make docker-windows-386'
-
-darwin-amd64: require-assets
+darwin: $(RESOURCES_DOT_GO)
 	@echo "Building darwin/amd64..." && \
 	export OSX_DEV_SDK=/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX$(OSX_MIN_VERSION).sdk && \
 	if [[ "$$(uname -s)" == "Darwin" ]]; then \
@@ -353,30 +327,7 @@ lantern: require-assets
 	$(call build-tags) && \
 	CGO_ENABLED=1 go build -race -o lantern -tags="$$BUILD_TAGS" -ldflags="$(LDFLAGS) $$EXTRA_LDFLAGS -s" github.com/getlantern/flashlight/main; \
 
-package-linux-386: require-version genassets linux-386
-	@echo "Generating distribution package for linux/386..." && \
-	$(call docker-up) && \
-	docker run -v $$PWD:/lantern -t $(DOCKER_IMAGE_TAG) /bin/bash -c 'cd /lantern && VERSION="'$$VERSION'" make docker-package-debian-386'
-
-package-linux-amd64: require-version genassets linux-amd64
-	@echo "Generating distribution package for linux/amd64..." && \
-	$(call docker-up) && \
-	docker run -v $$PWD:/lantern -t $(DOCKER_IMAGE_TAG) /bin/bash -c 'cd /lantern && VERSION="'$$VERSION'" make docker-package-debian-amd64'
-
-package-linux-arm: require-version genassets linux-arm
-	@echo "Generating distribution package for linux/arm..." && \
-	$(call docker-up) && \
-	docker run -v $$PWD:/lantern -t $(DOCKER_IMAGE_TAG) /bin/bash -c 'cd /lantern && VERSION="'$$VERSION'" HEADLESS="1" make docker-package-debian-arm'
-
 package-linux: require-version package-linux-386 package-linux-amd64
-
-package-windows: require-version windows
-	@echo "Generating distribution package for windows/386..." && \
-	if [[ -z "$$SECRETS_DIR" ]]; then echo "SECRETS_DIR environment value is required."; exit 1; fi && \
-	if [[ -z "$$BNS_CERT_PASS" ]]; then echo "BNS_CERT_PASS environment value is required."; exit 1; fi && \
-	$(call docker-up) && \
-	docker run -v $$PWD:/lantern -v $$SECRETS_DIR:/secrets -t $(DOCKER_IMAGE_TAG) /bin/bash -c 'cd /lantern && BNS_CERT="/secrets/bns.pfx" BNS_CERT_PASS="'$$BNS_CERT_PASS'" VERSION="'$$VERSION'" make docker-package-windows' && \
-	echo "-> lantern-installer.exe and lantern-installer-manoto.exe"
 
 package-darwin-manoto: require-version require-appdmg require-svgexport darwin
 	@echo "Generating distribution package for darwin/amd64 manoto..." && \
@@ -418,9 +369,13 @@ package-darwin: package-darwin-manoto
 		echo "-> Skipped: Can not generate a package on a non-OSX host."; \
 	fi;
 
-binaries: docker genassets linux windows darwin
+binaries: docker-assets docker-linux docker-windows darwin
 
-packages: require-version require-secrets clean update-dist binaries package-windows package-linux package-darwin
+packages: require-version require-secrets clean-desktop docker-package-windows docker-package-linux package-darwin
+
+# Override implicit docker-packages to avoid building whole packages target in
+# docker, since it builds the pieces it needs in docker itself.
+docker-packages: packages
 
 release-qa: require-version require-s3cmd
 	@BASE_NAME="lantern-installer-qa" && \
@@ -604,6 +559,27 @@ android-release: require-version require-secrets-dir $(LANTERN_MOBILE_ANDROID_RE
 android-install: $(LANTERN_MOBILE_ANDROID_DEBUG)
 	adb install -r $(LANTERN_MOBILE_ANDROID_DEBUG)
 
+clean-assets:
+	rm -f $(RESOURCES_DOT_GO)
+
+clean-bns-cert:
+	rm -f $(BNS_CERT)
+
+clean-desktop: clean-assets clean-bns-cert
+	rm -f lantern && \
+	rm -f lantern_linux* && \
+	rm -f lantern_darwin* && \
+	rm -f lantern_windows* && \
+	rm -f lantern-installer* && \
+	rm -f update_* && \
+	rm -f *.deb && \
+	rm -f *.png && \
+	rm -rf *.app && \
+	rm -f src/github.com/getlantern/flashlight/*.syso && \
+	rm -f *.dmg && \
+	rm -f $(LANTERN_MOBILE_TUN2SOCKS) && \
+	rm -rf $(LANTERN_MOBILE_DIR)/libs/armeabi*
+
 clean-mobile:
 	rm -f $(ANDROID_LIB) && \
 	rm -f $(ANDROID_SDK_ANDROID_LIB) && \
@@ -616,20 +592,8 @@ clean-mobile:
 	rm -f $(LANTERN_MOBILE_ANDROID_DEBUG) && \
 	rm -f $(LANTERN_MOBILE_ANDROID_RELEASE)
 
-clean: clean-mobile
-	rm -f lantern && \
-	rm -f lantern_linux* && \
-	rm -f lantern_darwin* && \
-	rm -f lantern_windows* && \
-	rm -f lantern-installer* && \
-	rm -f update_* && \
-	rm -f *.deb && \
-	rm -f *.png && \
-	rm -rf *.app && \
+clean-tooling:
 	rm -rf bin && \
-	rm -rf pkg && \
-	git checkout ./src/github.com/getlantern/flashlight/ui/resources.go && \
-	rm -f src/github.com/getlantern/flashlight/*.syso && \
-	rm -f *.dmg && \
-	rm -f $(LANTERN_MOBILE_TUN2SOCKS) && \
-	rm -rf $(LANTERN_MOBILE_DIR)/libs/armeabi*
+	rm -rf pkg
+
+clean: clean-tooling clean-desktop clean-mobile

--- a/Makefile
+++ b/Makefile
@@ -465,7 +465,7 @@ create-tag: require-version
 	@git tag -a "$$VERSION" -f --annotate -m"Tagged $$VERSION" && \
 	git push --tags -f
 
-test-and-cover:
+test-and-cover: $(RESOURCES_DOT_GO)
 	@echo "mode: count" > profile.cov && \
 	source setenv.bash && \
 	if [ -f envvars.bash ]; then \

--- a/Makefile
+++ b/Makefile
@@ -274,7 +274,7 @@ do-package-windows: require-version windows
 docker-package-windows: $(BNS_CERT)
 	make docker-do-package-windows
 
-linux: linux-386 linux-amd64
+linux: linux-386 linux-amd64 linux-arm
 
 system-checks:
 	@if [[ -z "$(DOCKER)" ]]; then echo 'Missing "docker" command.'; exit 1; fi && \
@@ -328,7 +328,7 @@ lantern: require-assets
 	$(call build-tags) && \
 	CGO_ENABLED=1 go build -race -o lantern -tags="$$BUILD_TAGS" -ldflags="$(LDFLAGS) $$EXTRA_LDFLAGS -s" github.com/getlantern/flashlight/main; \
 
-package-linux: require-version package-linux-386 package-linux-amd64
+package-linux: require-version package-linux-386 package-linux-amd64 package-linux-arm
 
 package-darwin-manoto: require-version require-appdmg require-svgexport darwin
 	@echo "Generating distribution package for darwin/amd64 manoto..." && \

--- a/Makefile
+++ b/Makefile
@@ -36,7 +36,14 @@ PACKAGED_YAML := .packaged-lantern.yaml
 MANOTO_YAML := .packaged-lantern-manoto.yaml
 
 RESOURCES_DOT_GO := ./src/github.com/getlantern/flashlight/ui/resources.go
-BNS_CERT := bns.pfx
+
+ifdef SECRETS_DIR
+BNS_CERT := $(SECRETS_DIR)/bns.pfx
+DOCKER_VOLS = "-v $$PWD:/lantern $(SECRETS_VOL) -v $$SECRETS_DIR:/secrets"
+else
+BNS_CERT := "../secrets/bns.pfx"
+DOCKER_VOLS = "-v $$PWD:/lantern $(SECRETS_VOL)"
+endif
 
 LANTERN_BINARIES_PATH ?= ../lantern-binaries
 
@@ -164,7 +171,8 @@ docker-%: system-checks
 	mkdir -p $$DOCKER_CONTEXT && \
 	cp Dockerfile $$DOCKER_CONTEXT && \
 	docker build -t $(DOCKER_IMAGE_TAG) $$DOCKER_CONTEXT && \
-	docker run -v $$PWD:/lantern -t $(DOCKER_IMAGE_TAG) /bin/bash -c 'cd /lantern && VERSION="'$$VERSION'" HEADLESS="'$$HEADLESS'" SECRETS_DIR="'$$SECRETS_DIR'" BNS_CERT_PASS="'$$BNS_CERT_PASS'" make $*';
+	echo docker run `echo $(DOCKER_VOLS) | xargs` -t $(DOCKER_IMAGE_TAG) /bin/bash -c 'cd /lantern && VERSION="'$$VERSION'" HEADLESS="'$$HEADLESS'" BNS_CERT_PASS="'$$BNS_CERT_PASS'" make $*' && \
+	docker run `echo $(DOCKER_VOLS) | xargs` -t $(DOCKER_IMAGE_TAG) /bin/bash -c 'cd /lantern && VERSION="'$$VERSION'" HEADLESS="'$$HEADLESS'" BNS_CERT_PASS="'$$BNS_CERT_PASS'" make $*';
 
 all: binaries
 android-dist: genconfig android
@@ -178,6 +186,7 @@ $(RESOURCES_DOT_GO): $(NPM)
 	git update-index --assume-unchanged $$APP/js/revision.js && \
 	DEST="$@" && \
 	cd $$LANTERN_UI && \
+	rm -rf node_modules && \
 	npm install && \
 	rm -Rf dist && \
 	gulp build && \
@@ -225,8 +234,8 @@ require-secrets-dir:
 require-secrets: require-secrets-dir
 	@if [[ -z "$$BNS_CERT_PASS" ]]; then echo "BNS_CERT_PASS environment value is required."; exit 1; fi
 
-$(BNS_CERT): require-secrets-dir
-	@cp -f $$SECRETS_DIR/$(BNS_CERT) $(BNS_CERT)
+$(BNS_CERT):
+	@if [[ ! -r "$(BNS_CERT)" ]]; then echo "Missing $(BNS_CERT)" && exit 1; fi
 
 require-lantern-binaries:
 	@if [[ ! -d "$(LANTERN_BINARIES_PATH)" ]]; then \
@@ -249,9 +258,7 @@ package-linux-arm: require-version linux-arm
 	@$(call fpm-debian-build,"arm")
 	@echo "-> lantern_$(VERSION)_arm.deb"
 
-package-windows: $(BNS_CERT) do-package-windows
-
-do-package-windows: require-version windows
+package-windows: $(BNS_CERT) require-version windows
 	@if [[ -z "$$BNS_CERT_PASS" ]]; then echo "BNS_CERT_PASS environment value is required."; exit 1; fi && \
 	INSTALLER_RESOURCES="installer-resources/windows" && \
 	rm -f $$INSTALLER_RESOURCES/$(PACKAGED_YAML) && \
@@ -269,12 +276,7 @@ do-package-windows: require-version windows
 	osslsigncode sign -pkcs12 "$(BNS_CERT)" -pass "$$BNS_CERT_PASS" -n "Lantern" -t http://timestamp.verisign.com/scripts/timstamp.dll -in "$$INSTALLER_RESOURCES/lantern-installer-unsigned.exe" -out "lantern-installer-manoto.exe" && \
 	echo "-> lantern-installer.exe and lantern-installer-manoto.exe"
 
-# We overwrite the implicit docker-package-windows rule to include a dependency
-# on $(BNS_CERT)
-docker-package-windows: $(BNS_CERT)
-	make docker-do-package-windows
-
-linux: linux-386 linux-amd64 linux-arm
+linux: linux-386 linux-amd64
 
 system-checks:
 	@if [[ -z "$(DOCKER)" ]]; then echo 'Missing "docker" command.'; exit 1; fi && \
@@ -328,7 +330,7 @@ lantern: require-assets
 	$(call build-tags) && \
 	CGO_ENABLED=1 go build -race -o lantern -tags="$$BUILD_TAGS" -ldflags="$(LDFLAGS) $$EXTRA_LDFLAGS -s" github.com/getlantern/flashlight/main; \
 
-package-linux: require-version package-linux-386 package-linux-amd64 package-linux-arm
+package-linux: require-version package-linux-386 package-linux-amd64
 
 package-darwin-manoto: require-version require-appdmg require-svgexport darwin
 	@echo "Generating distribution package for darwin/amd64 manoto..." && \
@@ -566,10 +568,7 @@ clean-assets:
 # Provided for backward compatibility with how people used to use the makefile
 update-dist: clean-assets assets
 
-clean-bns-cert:
-	rm -f $(BNS_CERT)
-
-clean-desktop: clean-assets clean-bns-cert
+clean-desktop: clean-assets
 	rm -f lantern && \
 	rm -f lantern_linux* && \
 	rm -f lantern_darwin* && \

--- a/Makefile
+++ b/Makefile
@@ -371,7 +371,7 @@ package-darwin: package-darwin-manoto
 
 binaries: docker-assets docker-linux docker-windows darwin
 
-packages: require-version require-secrets clean-desktop docker-package-windows docker-package-linux package-darwin
+packages: require-version require-secrets clean-desktop docker-assets docker-package-windows docker-package-linux package-darwin
 
 # Override implicit docker-packages to avoid building whole packages target in
 # docker, since it builds the pieces it needs in docker itself.

--- a/Makefile
+++ b/Makefile
@@ -204,6 +204,7 @@ linux-amd64: $(RESOURCES_DOT_GO)
 
 linux-arm: $(RESOURCES_DOT_GO)
 	@source setenv.bash && \
+	HEADLESS=1 && \
 	$(call build-tags) && \
 	CGO_ENABLED=1 CC=arm-linux-gnueabi-gcc CXX=arm-linux-gnueabi-g++ CGO_ENABLED=1 GOOS=linux GOARCH=arm GOARM=7 go build -a -o lantern_linux_arm -tags="$$BUILD_TAGS" -ldflags="$(LDFLAGS) $$EXTRA_LDFLAGS -linkmode internal -extldflags \"-static\"" github.com/getlantern/flashlight/main
 

--- a/README.md
+++ b/README.md
@@ -16,14 +16,20 @@
 ### Requisites
 
 * [Go 1.6rc1 or higher](https://golang.org/dl/).
-* [Docker](https://www.docker.com/).
 * [GNU Make](https://www.gnu.org/software/make/)
 * An OSX or Linux host.
+
+For release builds, you'll want to use Docker as well:
+
+* [Docker](https://www.docker.com/).
 
 We are going to create a Docker image that will take care of compiling Lantern
 for Windows and Linux, in order to compile Lantern for OSX you'll need an OSX
 host, this is a limitation caused by Lantern depending on C code and OSX build
 tools for certain features.
+
+Any target can be run on Docker by prefixing it with 'docker-', e.g.
+`make windows` runs locally and `make docker-windows` runs in docker.
 
 ### Docker Installation Instructions
 
@@ -173,7 +179,7 @@ The background image for the DMG is
 
 Lantern on Windows is distributed as an installer built with
 [nsis](http://nsis.sourceforge.net/). The installer is built and signed with
-`make package-windows`.
+`make docker-package-windows`.
 
 For `make package-windows` to be able to sign the executable, the environment variables
 `SECRETS_DIR` and `BNS_CERT_PASS` must be set to point to the secrets directory
@@ -193,7 +199,7 @@ Lantern on Ubuntu is distributed as a `.deb` package. You can generate a Debian
 package with:
 
 ```sh
-VERSION=2.0.0-beta2 make package-linux
+VERSION=2.0.0-beta2 make docker-package-linux
 ```
 
 The version string must match the Debian requirements:
@@ -204,8 +210,8 @@ This will build both 386 and amd64 packages.
 
 ### Generating all packages
 
-Use the `make packages` task combining all the arguments that `package-linux`,
-`package-windows` and `package-darwin` require.
+Use the `make packages` task combining all the arguments that
+`package-linux`, `package-windows` and `package-darwin` require.
 
 ```sh
 SECRETS_DIR=$PATH_TO_TOO_MANY_SECRETS BNS_CERT_PASS='***' \

--- a/src/github.com/getlantern/flashlight/ui/resources.go
+++ b/src/github.com/getlantern/flashlight/ui/resources.go
@@ -1,7 +1,0 @@
-// This file will be overwritten by make genassets.
-
-// +build !stub
-
-package ui
-
-var Resources = []byte{}

--- a/src/github.com/getlantern/lantern-ui/gulpfile.js
+++ b/src/github.com/getlantern/lantern-ui/gulpfile.js
@@ -1,3 +1,7 @@
+// es6-promise is required for script to work on older versions of nodejs like
+// what comes on old TravisCI build machines.
+require('es6-promise').polyfill();
+
 (function () {
   'use strict';
   var gulp = require('gulp');

--- a/src/github.com/getlantern/lantern-ui/package.json
+++ b/src/github.com/getlantern/lantern-ui/package.json
@@ -35,7 +35,8 @@
     "gulp-minify-html": "^1.0.3",
     "gulp-rev": "^5.0.1",
     "gulp-uglify": "^1.2.0",
-    "gulp-usemin": "^0.3.11"
+    "gulp-usemin": "^0.3.11",
+    "es6-promise": "~3.1.2"
   },
   "analyze": false,
   "license": "GPLv3",

